subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/DefaultLenientConfigurationTest.groovy

```diff
@@ -0,0 +1,135 @@
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.oldresult.TransientConfigurationResults
+import org.gradle.api.specs.Spec
+import org.gradle.internal.Factory
+import spock.lang.Specification
+
+class DefaultLenientConfigurationTest extends Specification {
+    def "should resolve first level dependencies in tree"() {
+        given:
+        TransientConfigurationResults transientConfigurationResults = Mock(TransientConfigurationResults)
+        DefaultLenientConfiguration lenientConfiguration = new DefaultLenientConfiguration(null, null, null, null, { transientConfigurationResults } as Factory)
+        ResolvedDependency root = Mock(ResolvedDependency)
+        def expectedResults = [Mock(ResolvedDependency)] as Set
+
+        when:
+        def results = lenientConfiguration.getFirstLevelModuleDependencies()
+
+        then:
+        1 * transientConfigurationResults.getRoot() >> root
+        1 * root.getChildren() >> expectedResults
+        results == expectedResults
+        0 * _._
+    }
+
+    def "should resolve and filter first level dependencies in tree"() {
+        given:
+        TransientConfigurationResults transientConfigurationResults = Mock(TransientConfigurationResults)
+        DefaultLenientConfiguration lenientConfiguration = new DefaultLenientConfiguration(null, null, null, null, { transientConfigurationResults } as Factory)
+        Spec spec = Mock(Spec)
+        def firstLevelDependencies = [(Mock(ModuleDependency)): Mock(ResolvedDependency), (Mock(ModuleDependency)): Mock(ResolvedDependency), (Mock(ModuleDependency)): Mock(ResolvedDependency)]
+        def firstLevelDependenciesEntries = firstLevelDependencies.entrySet() as List
+
+        when:
+        def result = lenientConfiguration.getFirstLevelModuleDependencies(spec)
+
+        then:
+        1 * transientConfigurationResults.getFirstLevelDependencies() >> firstLevelDependencies
+        1 * spec.isSatisfiedBy(firstLevelDependenciesEntries[0].key) >> true
+        1 * spec.isSatisfiedBy(firstLevelDependenciesEntries[1].key) >> false
+        1 * spec.isSatisfiedBy(firstLevelDependenciesEntries[2].key) >> true
+        result == [firstLevelDependenciesEntries[0].value, firstLevelDependenciesEntries[2].value] as Set
+        0 * _._
+    }
+
+    def "should flatten all resolved dependencies in dependency tree"() {
+        given:
+        TransientConfigurationResults transientConfigurationResults = Mock(TransientConfigurationResults)
+        DefaultLenientConfiguration lenientConfiguration = new DefaultLenientConfiguration(null, null, null, null, { transientConfigurationResults } as Factory)
+
+        def (expected, root) = generateDependenciesWithChildren(treeStructure)
+
+        when:
+        def result = lenientConfiguration.allDependencies()
+
+        then:
+        1 * transientConfigurationResults.getRoot() >> root
+        result.size() == size
+        result.size() == expected.size()
```

Comment:
Assertion probably not necessary as we are comparing expected result with the actual result.